src/Interpreters/AsynchronousMetrics.cpp

```diff
@@ -674,21 +700,32 @@ void AsynchronousMetrics::update(std::chrono::system_clock::time_point update_ti
         {
             Int64 amount = total_memory_tracker.get();
             Int64 peak = total_memory_tracker.getPeak();
-            Int64 new_amount = data.resident;
+            Int64 rss = data.resident;
 
-            Int64 difference = new_amount - amount;
+#if USE_JEMALLOC
+            /// This is a memory which is kept by allocator.
+            /// Remove it from RSS to decrease memory drift.
+            rss -= je_malloc_mapped - je_malloc_allocated;
```

Comment:
But if it will not take into account this memory, then the drift should be higher, since `MemoryTracker` knows about some of them (`max_untracker_memory`).
What did I miss?

IIUC then we should substract retained memory, then it should be [`stats.retained`](https://jemalloc.net/jemalloc.3.html#stats.retained), and actually to resolve issues like in #39614, it should not be take into account.

Reply:
> the drift should be higher

I did not understand the idea why.
My idea is that `je_malloc_mapped - je_malloc_allocated` is an amount of memory which is hold by jemalloc arenas, and can be used for father allocations. Basically this is the difference between RSS and MemoryTracking I see (for simply group by queries).

`max_untracker_memory` is a memory which is accounted in `je_malloc_allocated` already (but maybe not by MemoryTracker yet). It may only make a check vs RSS stricter (cause during the next second max_untracker_memory can be accounted twice). But it is the same thing which we had in previous approach.

> IIUC then we should substract retained memory

Retained memory is a memory which was marked as `MADV_DONTNEED ` or `MADV_FREE` (and likely la latter one). In case `MADV_DONTNEED` everything is fine cause those pages are accounted in RSS immediately (decreasing it). For `MADV_FREE` pages everything is harder cause in general you can't tell how many pages were actually freed by system (and accounted in RSS) and which were not. But actually what I seen (for simple group by queries) that retained pages are taken by system pretty fast (and the main reason fro drifts is memory from je_malloc_mapped).

Also, allocated + retained may be much bigger than RSS from system.

> FWIW I've tried to run binary from this PR, and it reports lots of such messages for simple query with group by.

last build worked ok to me. May I ask you to give me an example query?

Reply:
>My idea is that je_malloc_mapped - je_malloc_allocated is an amount of memory which is hold by jemalloc arenas, and can be used for father allocations

Egh, I've interpreted this incorrectly. Sorry for the noise.

>May I ask you to give me an example query?

Query:

```sql
select * from numbers_mt(1e8) group by number format Null
```

But looking at this deeper, this looks OK.